nntrainer/layers/mol_attention_layer.cpp

```diff
@@ -24,18 +25,103 @@ MoLAttentionLayer::~MoLAttentionLayer() {}
 
 static constexpr size_t SINGLE_INOUT_IDX = 0;
 
-enum AttentionParams { query = 0, value = 1, key = 2, score, weights };
+enum AttentionParams { query = 0, value = 1, state = 2, mlp_w, mlp_proj_w };
 
 void MoLAttentionLayer::finalize(InitLayerContext &context) {
-  finalizeCommon(context);
-  /** NYI */
+  if (context.getNumInputs() != 3)
+    throw std::runtime_error("MoL Attention layer needs 3 inputs.");
+
+  auto const &all_dims = context.getInputDimensions();
+  auto const &query_dim = all_dims[AttentionParams::query];
+
+  wt_idx[AttentionParams::query] = AttentionParams::query;
+  wt_idx[AttentionParams::value] = AttentionParams::value;
+  wt_idx[AttentionParams::state] = AttentionParams::state;
+
+  softmax.setActiFunc(ActivationType::ACT_SOFTMAX);
+  tanh.setActiFunc(ActivationType::ACT_TANH);
+  sigmoid.setActiFunc(ActivationType::ACT_SIGMOID);
+
+  auto unit = std::get<props::Unit>(mol_props).get();
+  auto mol_k = std::get<props::MoL_K>(mol_props).get();
+
+  TensorDim mlp_w_dim = {query_dim.width(), unit};
+  wt_idx[AttentionParams::mlp_w] =
+    context.requestWeight(mlp_w_dim, Tensor::Initializer::XAVIER_UNIFORM,
+                          WeightRegularizer::NONE, 0.0, "mlp_w", true);
+
+  TensorDim mlp_proj_w_dim = {unit, 3 * mol_k};
+  wt_idx[AttentionParams::mlp_proj_w] =
+    context.requestWeight(mlp_proj_w_dim, Tensor::Initializer::XAVIER_UNIFORM,
+                          WeightRegularizer::NONE, 0.0, "mlp_proj_w", true);
+
+  context.setOutputDimensions({query_dim});
 }
 
 void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
-  /** NYI */
+  Tensor &query = context.getInput(wt_idx[AttentionParams::query]);
+  Tensor &value = context.getInput(wt_idx[AttentionParams::value]);
+  Tensor &state = context.getInput(wt_idx[AttentionParams::state]);
+
+  Tensor &output = context.getOutput(SINGLE_INOUT_IDX);
+  Tensor &mlp_w = context.getWeight(wt_idx[AttentionParams::mlp_w]);
+  Tensor &mlp_proj_w = context.getWeight(wt_idx[AttentionParams::mlp_proj_w]);
+
+  const TensorDim &input_dim = query.getDim();
+  unsigned int batch = input_dim.batch();
+  auto mol_k = std::get<props::MoL_K>(mol_props).get();
+
+  Tensor mlp_out = query.dot(mlp_w);
+
+  Tensor mlp_tanh;
+  tanh.run_fn(mlp_out, mlp_tanh);
+
+  Tensor mlp_proj_out = mlp_tanh.dot(mlp_proj_w);
+
+  /** @note kappa_hat, beta_hat, alpha_hat are strided */
+  Tensor kappa_hat = mlp_proj_out.getSharedDataTensor({batch, mol_k}, 0, false);
+  Tensor beta_hat =
+    mlp_proj_out.getSharedDataTensor({batch, mol_k}, mol_k, false);
+  Tensor alpha_hat =
+    mlp_proj_out.getSharedDataTensor({batch, mol_k}, mol_k * 2, false);
+
+  Tensor kappa = kappa_hat.apply(static_cast<float (*)(float)>(&std::exp));
+  Tensor beta = beta_hat.apply(static_cast<float (*)(float)>(&std::exp));
```

Comment:
I guess we can use std::expf to avoid casting would happen inside function?

Reply:
Will use expf (not std::expf) in upcoming PR.